### PR TITLE
Cancel amqp token refresher as a part of transport cleanup

### PR DESCRIPTION
**Pull request by abhipsaMisra**
The amqp sas token refresher runs as a background unmonitored task. We need to ensure that the task is cancelled as a part of transport layer cleanup.

Related: #2339 , maybe #2351

### CODE REVIEW

**Review by patilsnr — APPROVED** [review]

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:132]: Protect against null?

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:56]: Is this method re-entrant? If so, should we dispose before overwriting?

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:56]: The current implementation has upstream caller protection against multiple threads calling it in parallel. However, this method is defined in an interface, so I'll update it to check for an existing source and reuse it if it exists. I can't think of a scenario where this method would be called by concurrently with different cancellation tokens, so us reusing the previous token source should be fine.

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (abhipsaMisra) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:57]: the original cancellation token passed into this API depends on the cancellation token of the upstream call, i.e. the cancellation token passed into OpenAsync() etc. I'd like to have that decoupled since the token refresher should ideally run on a separate thread all the way until it is canceled.

**Review by patilsnr — APPROVED** [review]

**Comment by patilsnr** [comment]
Optimistic sign off, would recommend waiting for another approval from someone more experienced.

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:54]: Can we detect if it isn't and assert?

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:89]:
```suggestion
                    Logging.Info(this, refreshesOn, $"{_amqpIotCbsTokenProvider} before {nameof(RefreshLoopAsync)}");
```

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:121]: Braces can be removed.

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:89]: Yeah. The instance hash was useful to me when debugging multiplexed devices, so I thought it would be useful to add it into the log.

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:166]: I can remove this, sure

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**
